h2/src/main/org/h2/expression/Comparison.java

```diff
@@ -200,7 +203,7 @@ public Expression optimize(Session session) {
                     // If not the column values will need to be promoted
                     // to constant type, but vise versa, then let's do this here once.
                     if (constType != resType)
-                        right = ValueExpression.get(r.convertTo(resType));
+                        right = ValueExpression.get(((ExpressionColumn) left).getColumn().convert(r));
```

Comment:
I don't really see why. Before this change, `Column`'s `convert` did exactly the same thing as the old  conversion code here (and could have been used right away). Please see my thoughts about binding of the value type to column below. After all, here we clearly want to convert constant to type of the column, so asking the column to do the conversion seems rational.